File: packages/sim/src/ledger.ts
/** RI-01: resource conservation — the accounting check across production, transfer, delivery commitment and
 *  consumption (`PROGRESS.md` RI-01; the revised plan's RI-01 row). Every item the tile layer knows (flow.ts `ITEMS`)
 *  is counted where it is now (`held`: the chest, the line buffer, the ring's stand-in hoppers, the pockets, the
 *  machines' inventories and the belts) against what has entered the game since the flow layer opened (rubble units
 *  mined and recipe output, `sources`) and what has left it (recipe inputs, machine prices, claim prices, repairs,
 *  burned coal, fired rounds and rounds a full buffer could not take, `sinks`). Transfers — belts, inserters, the
 *  chest, hand-feeds, the ring's draw from the line buffer, kits — move an item between places and appear on neither
 *  side. `unexplained` is held + sinks − sources by item, against the opening stock: zero when every unit is
 *  explained; positive when items appeared from nowhere, negative when they vanished. Rubble still in the ground is
 *  not an item and is not counted; `stats.minedOf` is where it enters.
 *
 *  The opening stock is recorded on the flow state when it is created (`FlowState.ledger`), as held − net at that
 *  moment, so a save from before RI-01 is checked from the moment it was upgraded rather than from a history it never
 *  kept (flow.ts `upgrade`). Kits are free (chestCount 'kit' is Infinity) and machines carried in the pockets are
 *  not items: their price left the game at their first placement (`stats.placed`). */
import { SimState } from './types';
import { ROUNDS_PER_MAG } from './constants';
import { Item, ITEMS, zeroItems, isItem, recipeOf, recipeOutput, REPAIR_COPPER, ensureFlow } from './flow';   // flow.ts imports openLedger back: only functions cross, at call time

/** Where the items are, one column a place. */
/** RI-03: `committed` is claim material delivered to a Dark block's installation and not yet activated (plan §4.3:
 *  in an inventory, committed to a restoration stage, or consumed — never two of them). */
export type LedgerPlace = 'chest' | 'buffer' | 'ring' | 'pockets' | 'machines' | 'belts' | 'committed';
export const LEDGER_PLACES: readonly LedgerPlace[] = ['chest', 'buffer', 'ring', 'pockets', 'machines', 'belts', 'committed'];

export interface Ledger {
  ok: boolean;
  tolerance: number;
  /** The flow tick the opening stock was recorded at (0 on a city played from its start). */
  openedAt: number;
  opening: Record<Item, number>;
  sources: Record<Item, number>;
  sinks: Record<Item, number>;
  held: Record<Item, number>;
  where: Record<LedgerPlace, Record<Item, number>>;
  unexplained: Record<Item, number>;
  /** One line an item out of tolerance. */
  problems: string[];
}

const ROUNDS = ROUNDS_PER_MAG;

/** The flow state as it stands — `ensureFlow` itself calls `openLedger`, so the counters never call back into it. */
function flowOf(st: SimState) { const f = st.flow; if (!f) throw new Error('ledger: the flow layer is not open (call ensureFlow first)'); return f; }

/** Every item where it stands now, by place and in total. Rounds count as magazines (ten a magazine, fractional). */
export function heldItems(st: SimState): { total: Record<Item, number>; where: Record<LedgerPlace, Record<Item, number>> } {
  const f = flowOf(st);
  const where = {} as Record<LedgerPlace, Record<Item, number>>;
  for (const p of LEDGER_PLACES) where[p] = zeroItems();
  const add = (p: LedgerPlace, k: string, n: number) => { if (isItem(k) && n > 0) where[p][k] += n; };
  add('chest', 'steel', st.stock.steel); add('chest', 'copper', st.stock.copper); add('chest', 'stone', st.stock.stone);
  for (const k in f.store) add('chest', k, f.store[k as keyof typeof f.store]);
  add('buffer', 'magazine', st.buffer / ROUNDS);
  for (const e of st.ring) if (!e.turrets) add('ring', 'magazine', e.hopper / ROUNDS);   // a stand-in edge's hopper; a turret edge's mirrors its turrets
  for (const k in st.engineer.inv) add('pockets', k, st.engineer.inv[k]);
  for (const bi in f.delivered ?? {}) { add('committed', 'steel', f.delivered[bi].steel); add('committed', 'copper', f.delivered[bi].copper); }
  for (const site of Object.values(st.campaign?.expansion ? { station: st.campaign.expansion.station, radio: st.campaign.expansion.radio } : {})) { add('committed', 'steel', site.delivered.steel); add('committed', 'copper', site.delivered.copper); }
  for (const cb of f.heart?.cabinets ?? []) { add('committed', 'steel', cb.delivered.steel); add('committed', 'copper', cb.delivered.copper); }   // RI-06: the feeder cabinets' materials until the Heart is destroyed
  for (const m of f.machines) {
    for (const it of m.items) add('belts', it.k, 1);
    if (m.hold) add(m.kind === 'belt' ? 'belts' : 'machines', m.hold, 1);
    if (m.kind === 'turret') add('machines', 'magazine', (m.inv.rounds ?? 0) / ROUNDS);
    else { for (const k in m.inv) add('machines', k, m.inv[k]); if (m.out > 0) add('machines', m.kind === 'assembler' ? recipeOutput(recipeOf(m)) : 'magazine', m.out); }
    if (m.cargo) for (const k in m.cargo) add('machines', k, m.cargo[k]);   // RI-05: a tram's load, a stop's arrivals
  }
  const total = zeroItems();
  for (const p of LEDGER_PLACES) for (const k of ITEMS) total[k] += where[p][k];
  return { total, where };
}

/** What the counters say entered the game (`sources`) and left it (`sinks`) since the flow layer's counters began. */
export function ledgerFlows(st: SimState): { sources: Record<Item, number>; sinks: Record<Item, number> } {
  const f = flowOf(st), s = f.stats, b = st.stats;
  const sources = zeroItems(), sinks = zeroItems();
  for (const k of ITEMS) { sources[k] += (s.minedOf[k] ?? 0) + (s.made[k] ?? 0); sinks[k] += s.consumed[k] ?? 0; }
  sinks.steel += (s.placed?.steel ?? 0) + (b.spentSteel ?? 0);
  sinks.copper += (s.placed?.copper ?? 0) + (b.spentCopper ?? 0) + (f.repairs ?? 0) * REPAIR_COPPER;
  sinks.coal += s.coalBurned ?? 0;
  sinks.magazine += ((s.fired ?? 0) + (st.engineer.fired ?? 0) + (b.ringFired ?? 0) + (b.roundsLost ?? 0)) / ROUNDS;
  return { sources, sinks };
}

/** The opening record for a flow state: held − net now, so that `conservation` reads zero from here on. */
export function openLedger(st: SimState): { tick: number; base: Record<Item, number> } {
  const { total } = heldItems(st), { sources, sinks } = ledgerFlows(st);
  const base = zeroItems();
  for (const k of ITEMS) base[k] = total[k] - (sources[k] - sinks[k]);
  return { tick: st.flow?.tick ?? 0, base };
}

/** The conservation check. `tolerance` absorbs floating-point drift in fractional rounds and coal (a hundredth of an item). */
export function conservation(st: SimState, tolerance = 0.01): Ledger {
  const f = ensureFlow(st);
  const { total: held, where } = heldItems(st), { sources, sinks } = ledgerFlows(st);
  const opening = f.ledger?.base ?? zeroItems(), openedAt = f.ledger?.tick ?? 0;
  const unexplained = zeroItems(), problems: string[] = [];
  for (const k of ITEMS) {
    unexplained[k] = held[k] + sinks[k] - sources[k] - opening[k];
    if (Math.abs(unexplained[k]) > tolerance) problems.push(`${k}: ${unexplained[k] > 0 ? '+' : ''}${unexplained[k].toFixed(2)} unexplained (opening ${opening[k].toFixed(1)} + sources ${sources[k].toFixed(1)} = held ${held[k].toFixed(1)} + sinks ${sinks[k].toFixed(1)})`);
  }
  return { ok: problems.length === 0, tolerance, openedAt, opening, sources, sinks, held, where, unexplained, problems };
}

File: Unity/Relight/Assets/Relight/Sim/Actor/Ledger.cs
using System;
using System.Collections.Generic;
using System.Globalization;

namespace Relight.Sim
{
    /// <summary>
    /// Where the ledger counts an item (reference ledger.ts `LedgerPlace`).
    /// Phase B fills <see cref="Pockets"/> and <see cref="Machines"/> only.
    /// <see cref="Belts"/> (belt contents, an inserter's hand, a tram's cargo) and <see cref="Committed"/>
    /// (materials delivered to a restoration site and not yet activated) are Phase C and are always zero here.
    /// The reference's `chest` (the Depot's abstract stock) and `buffer` (the line buffer) places, and its `ring`
    /// place, belong to the retired block/lattice economy (U-D-32) and are not ported: the physical supply chest
    /// is a machine and counts under <see cref="Machines"/>.
    /// </summary>
    public enum LedgerPlace
    {
        Pockets = 0,
        Machines = 1,
        Belts = 2,
        Committed = 3,
    }

    /// <summary>
    /// RI-01 / TA §2.7 / U-D-05 conservation: every item is counted where it is now (`held`) against what entered
    /// the game (`sources`) and left it (`sinks`) since the ledger opened.
    /// <c>unexplained[k] = held[k] + sinks[k] − sources[k] − opening[k]</c>; zero means every unit is explained.
    /// Ported from reference ledger.ts (`heldItems`, `ledgerFlows`, `openLedger`, `conservation`).
    /// One round is one bullet (U-D-08), so the reference's ROUNDS_PER_MAG divisions collapse to 1.
    /// </summary>
    public static class Ledger
    {
        public const int PlaceCount = 4;
        /// <summary>Reference flow.ts REPAIR_COPPER.</summary>
        public const double RepairCopper = 1;

        /// <summary>Reference `heldItems`: every item where it stands now, by place and in total.</summary>
        public static void Held(SimState st, GameData d, ItemCounts[] where, ItemCounts total)
        {
            for (var p = 0; p < PlaceCount; p++) where[p].Clear();
            total.Clear();

            var pockets = where[(int)LedgerPlace.Pockets];
            var inv = st.Engineer.Inv;
            for (var i = 0; i < Items.Count; i++)
            {
                var n = inv[(ItemId)i];
                if (n > 0) pockets.Add((ItemId)i, n);
            }
            // Phase C equipment: a holstered weapon's `loaded` rounds count as magazines in the pockets.
            // Phase B has no equipment, so that term is fixed at zero.

            // A cancelled hand-craft batch's plates that the Backpack could not take back yet are still the
            // engineer's: they are owed to the pockets and HandCraft.Drain pays them in as space appears, so they
            // are counted here rather than left as a sink. No new LedgerPlace — the four slots are unchanged.
            var hand = st.Hand;
            if (hand != null)
            {
                if (hand.RefundSteel > 0) pockets.Add(ItemId.Steel, hand.RefundSteel);
                if (hand.RefundCopper > 0) pockets.Add(ItemId.Copper, hand.RefundCopper);
            }

            var machines = where[(int)LedgerPlace.Machines];
            for (var mi = 0; mi < st.Machines.Count; mi++)
            {
                var m = st.Machines[mi];
                if (string.Equals(m.Kind, "turret", StringComparison.Ordinal))
                {
                    if (m.Rounds > 0) machines.Add(ItemId.Magazine, m.Rounds);
                    continue;
                }
                for (var i = 0; i < Items.Count; i++)
                {
                    var n = m.Inv[(ItemId)i];
                    if (n > 0) machines.Add((ItemId)i, n);
                }
                // Reference: a processor's finished output is its recipe's output item, anything else is magazines.
                // Phase B machines carry no recipe (production is Phase C), so `out` counts as magazines throughout.
                if (m.Out > 0) machines.Add(ItemId.Magazine, m.Out);
            }

            for (var p = 0; p < PlaceCount; p++)
                for (var i = 0; i < Items.Count; i++) total.Add((ItemId)i, where[p][(ItemId)i]);
        }

        /// <summary>Reference `ledgerFlows`: what the counters say entered and left the game.</summary>
        public static void Flows(SimState st, ItemCounts sources, ItemCounts sinks)
        {
            sources.Clear();
            sinks.Clear();
            var s = st.Stats;
            for (var i = 0; i < Items.Count; i++)
            {
                var k = (ItemId)i;
                sources.Add(k, s.MinedOf[k] + s.Made[k]);
                sinks.Add(k, s.Consumed[k] + s.Placed[k]);
            }
            sinks.Add(ItemId.Steel, s.SpentSteel);
            sinks.Add(ItemId.Copper, s.SpentCopper + s.Repairs * RepairCopper);
            sinks.Add(ItemId.Coal, s.CoalBurned);
            sinks.Add(ItemId.Magazine, s.Fired + s.EngineerFired + s.RoundsLost);
        }

        /// <summary>Reference `openLedger`: held − net now, so <see cref="Conservation"/> reads zero from here on.</summary>
        public static LedgerOpening Open(SimState st, GameData d)
        {
            var where = NewPlaces();
            var total = new ItemCounts();
            var sources = new ItemCounts();
            var sinks = new ItemCounts();
            Held(st, d, where, total);
            Flows(st, sources, sinks);
            var opening = new LedgerOpening { Tick = st.Tick };
            for (var i = 0; i < Items.Count; i++)
            {
                var k = (ItemId)i;
                opening.Base[k] = total[k] - (sources[k] - sinks[k]);
            }
            return opening;
        }

        /// <summary>Reference `conservation`. `tolerance` absorbs floating-point drift (a hundredth of an item).</summary>
        public static LedgerView Conservation(SimState st, GameData d, double tolerance = 0.01)
        {
            var where = NewPlaces();
            var held = new ItemCounts();
            var sources = new ItemCounts();
            var sinks = new ItemCounts();
            Held(st, d, where, held);
            Flows(st, sources, sinks);
            var opening = st.Ledger?.Base?.Clone() ?? new ItemCounts();
            var openedAt = st.Ledger?.Tick ?? 0;
            var unexplained = new ItemCounts();
            var problems = new List<string>();
            var inv = CultureInfo.InvariantCulture;
            for (var i = 0; i < Items.Count; i++)
            {
                var k = (ItemId)i;
                var v = held[k] + sinks[k] - sources[k] - opening[k];
                unexplained[k] = v;
                if (Math.Abs(v) <= tolerance) continue;
                problems.Add($"{Items.Key(k)}: {(v > 0 ? "+" : "")}{v.ToString("F2", inv)} unexplained " +
                             $"(opening {opening[k].ToString("F1", inv)} + sources {sources[k].ToString("F1", inv)} = " +
                             $"held {held[k].ToString("F1", inv)} + sinks {sinks[k].ToString("F1", inv)})");
            }
            return new LedgerView(problems.Count == 0, tolerance, openedAt, opening, sources, sinks, held, where, unexplained, problems);
        }

        public static ItemCounts[] NewPlaces()
        {
            var a = new ItemCounts[PlaceCount];
            for (var i = 0; i < PlaceCount; i++) a[i] = new ItemCounts();
            return a;
        }
    }
}
